File: android/app/src/main/kotlin/dev/agentdeck/terrarium/creature/CloudCreature.kt
package dev.agentdeck.terrarium.creature

import android.graphics.Paint
import android.graphics.Typeface
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathFillType
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.DrawScope
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.withTransform
import androidx.compose.ui.graphics.nativeCanvas
import androidx.compose.ui.graphics.vector.PathParser
import dev.agentdeck.terrarium.CreatureGeometry
import dev.agentdeck.terrarium.CreatureNameTagStyle
import dev.agentdeck.terrarium.creatureNameTagMetric
import dev.agentdeck.terrarium.normalizeSvgArcFlags
import dev.agentdeck.terrarium.resolveCreatureNameTagLayout
import dev.agentdeck.terrarium.OctopusVisualState
import dev.agentdeck.terrarium.TerrariumColors
import dev.agentdeck.terrarium.TerrariumLayout
import dev.agentdeck.terrarium.TerrariumTiming
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

/**
 * Codex CLI creature using the exact design/brand/codex.svg geometry with a
 * blue-purple gradient. The official path contains the `>_` cutout.
 *
 * Colors: lavender (#B394E5) top → deep blue (#3342C7) bottom gradient.
 * Working state shows bioluminescent orbiting particles.
 * Same public API as OctopusCreature for interchangeable use.
 */
class CloudCreature(
    centerXFraction: Float = TerrariumLayout.OCTOPUS_CENTER_X_FRACTION,
    centerYFraction: Float = TerrariumLayout.OCTOPUS_CENTER_Y_FRACTION,
    private var scaleFactor: Float = 1f,
    phaseOffset: Float = 0f,
    displayName: String? = null,
) : Creature {

    private var visualState by mutableStateOf(OctopusVisualState.FLOATING)
    private var time by mutableFloatStateOf(phaseOffset)
    private var transitionProgress by mutableFloatStateOf(1f)
    private var nameTag: String? by mutableStateOf(displayName)
    private var showNameTag by mutableStateOf(displayName != null)

    // Swimming state
    private var homeX = centerXFraction
    private var homeY = centerYFraction
    private var currentX = centerXFraction
    private var currentY = centerYFraction
    private var targetX = centerXFraction
    private var targetY = centerYFraction
    private var waypointTimer = 0f
    private var waypointInterval = TerrariumTiming.WAYPOINT_MIN_INTERVAL +
        kotlin.random.Random.nextFloat() * (TerrariumTiming.WAYPOINT_MAX_INTERVAL - TerrariumTiming.WAYPOINT_MIN_INTERVAL)
    private val standingJitter = kotlin.random.Random.nextFloat() * 0.06f - 0.03f

    // Bioluminescent particle state (8 orbiting particles when WORKING)
    private data class GlowParticle(
        val orbitRadius: Float,
        val orbitSpeed: Float,
        val phaseOffset: Float,
        val size: Float,
        val color: Color,
    )

    private val glowParticles = List(GLOW_PARTICLE_COUNT) { i ->
        GlowParticle(
            orbitRadius = 0.7f + kotlin.random.Random.nextFloat() * 0.6f, // 0.7–1.3 of body radius
            orbitSpeed = 0.8f + kotlin.random.Random.nextFloat() * 0.6f,
            phaseOffset = i * (2f * PI.toFloat() / GLOW_PARTICLE_COUNT),
            size = 0.04f + kotlin.random.Random.nextFloat() * 0.03f,
            color = if (i % 2 == 0) GLOW_CYAN else GLOW_LAVENDER,
        )
    }

    /** Callback invoked when transitioning away from ASKING (bubble pop trigger). */
    var onAskingExit: ((nx: Float, ny: Float) -> Unit)? = null

    fun setState(newState: OctopusVisualState) {
        if (newState != visualState) {
            if (visualState == OctopusVisualState.ASKING) {
                onAskingExit?.invoke(currentX, currentY)
            }
            visualState = newState
            transitionProgress = 0f
        }
    }

    fun setDisplayName(name: String?, show: Boolean = name != null) {
        nameTag = name
        showNameTag = show
    }

    /** Update home position — creature lerps naturally (no teleport). */
    fun setHomePosition(x: Float, y: Float, scale: Float) {
        homeX = x
        homeY = y
        scaleFactor = scale
    }

    /** Current live position for tetra attractor tracking. */
    fun currentPosition(): Pair<Float, Float> = currentX to currentY

    /** Whether this cloud is currently working (swimming, scattering data). */
    fun isWorking(): Boolean = visualState == OctopusVisualState.WORKING

    override fun update(dt: Float) {
        time += dt
        if (transitionProgress < 1f) {
            transitionProgress = (transitionProgress + dt * 3f).coerceAtMost(1f)
        }

        when (visualState) {
            OctopusVisualState.SLEEPING -> {
                val myDeepY = STANDING_Y_DEEP + standingJitter * 0.5f
                currentX += (homeX - currentX) * dt * 4f
                currentY += (myDeepY - currentY) * dt * 4f
            }
            OctopusVisualState.FLOATING -> {
                val myStandingY = (STANDING_Y + standingJitter + (homeX - 0.4f) * 0.15f).coerceAtMost(0.65f)
                val breathBob = sin(time * 0.6f) * 0.003f
                val idleSway = sin(time * 0.25f) * 0.004f
                currentX += (homeX + idleSway - currentX) * dt * 4f
                currentY += (myStandingY + breathBob - currentY) * dt * 4f
            }
            OctopusVisualState.ASKING -> {
                val myStandingY = (ASKING_Y + standingJitter + (homeX - 0.4f) * 0.15f).coerceAtMost(0.65f)
                val fidgetX = sin(time * 1.0f) * 0.006f
                currentX += (homeX + fidgetX - currentX) * dt * 4f
                currentY += (myStandingY - currentY) * dt * 4f
            }
            OctopusVisualState.WORKING -> {
                val lane = swimLane()
                waypointTimer += dt
                if (waypointTimer >= waypointInterval) {
                    waypointTimer = 0f
                    waypointInterval = TerrariumTiming.WAYPOINT_MIN_INTERVAL +
                        kotlin.random.Random.nextFloat() * (TerrariumTiming.WAYPOINT_MAX_INTERVAL - TerrariumTiming.WAYPOINT_MIN_INTERVAL)
                    pickNewWaypoint()
                }
                val rate = TerrariumTiming.SWIM_LERP_RATE * dt
                currentX += (targetX - currentX) * rate
                currentY += (targetY - currentY) * rate
                currentX = currentX.coerceIn(lane.minX, lane.maxX)
                currentY = currentY.coerceIn(lane.minY, lane.maxY)
            }
        }
    }

    private fun pickNewWaypoint() {
        val lane = swimLane()
        val angle = kotlin.random.Random.nextFloat() * 2f * PI.toFloat()
        val radiusX = maxOf(0.07f, (lane.maxX - lane.minX) * 0.48f)
        val radiusY = maxOf(0.04f, (lane.maxY - lane.minY) * 0.40f)
        targetX = (lane.centerX + cos(angle) * radiusX).coerceIn(lane.minX, lane.maxX)
        targetY = (lane.centerY + sin(angle) * radiusY).coerceIn(lane.minY, lane.maxY)
    }

    private data class SwimLane(
        val minX: Float,
        val maxX: Float,
        val minY: Float,
        val maxY: Float,
        val centerX: Float,
        val centerY: Float,
    )

    private fun swimLane(): SwimLane {
        val halfWidth = minOf(0.16f, maxOf(0.09f, 0.08f + scaleFactor * 0.05f))
        val centerX = homeX.coerceIn(TerrariumLayout.SWIM_MIN_X + 0.06f, TerrariumLayout.SWIM_MAX_X - 0.06f)
        val centerY = homeY.coerceIn(WORKING_MIN_Y + 0.04f, WORKING_MAX_Y - 0.04f)
        val verticalSlack = minOf(0.08f, maxOf(0.04f, 0.04f + scaleFactor * 0.02f))
        return SwimLane(
            minX = maxOf(TerrariumLayout.SWIM_MIN_X, centerX - halfWidth),
            maxX = minOf(TerrariumLayout.SWIM_MAX_X, centerX + halfWidth),
            minY = maxOf(WORKING_MIN_Y, centerY - verticalSlack),
            maxY = minOf(WORKING_MAX_Y, centerY + verticalSlack),
            centerX = centerX,
            centerY = centerY,
        )
    }

    override fun draw(scope: DrawScope) {
        val w = scope.size.width
        val h = scope.size.height
        val baseWidth = minOf(w, h * 2f)

        val bodyRadius = baseWidth * TerrariumLayout.OCTOPUS_BODY_RADIUS_FRACTION * scaleFactor
        val centerX = w * currentX

        // Bob animation
        val bobOffset = when (visualState) {
            OctopusVisualState.WORKING -> sin(time * 2f * PI.toFloat() / (TerrariumTiming.FLOAT_PERIOD_MS / 1000f)) *
                h * TerrariumTiming.FLOAT_AMPLITUDE_FRACTION
            OctopusVisualState.FLOATING -> sin(time * 0.6f) * h * 0.003f
            else -> 0f
        }
        val centerY = h * currentY + bobOffset

        val bodyAlpha = when (visualState) {
            OctopusVisualState.SLEEPING -> 0.35f
            else -> 1f
        }

        // Bioluminescent glow particles (behind body, WORKING only)
        if (visualState == OctopusVisualState.WORKING) {
            drawGlowParticles(scope, centerX, centerY, bodyRadius, bodyAlpha)
        }

        // Draw canonical Codex mark
        drawCloudBody(scope, centerX, centerY, bodyRadius, bodyAlpha)

        // ASKING: speech bubble with "?"
        if (visualState == OctopusVisualState.ASKING) {
            drawSpeechBubble(scope, centerX, centerY, bodyRadius)
        }

        // Name tag
        if (showNameTag && nameTag != null) {
            drawNameTag(scope, centerX, centerY, bodyRadius, nameTag!!)
        }
    }

    /** Draw the canonical Codex path; animation changes scale/color only. */
    private fun drawCloudBody(
        scope: DrawScope,
        cx: Float, cy: Float,
        bodyRadius: Float,
        alpha: Float,
    ) {
        val breathScale = when (visualState) {
            OctopusVisualState.WORKING -> 1f + sin(time * TerrariumTiming.THINKING_PULSE_SPEED) * 0.04f
            OctopusVisualState.FLOATING -> 1f + sin(time * 0.8f) * 0.015f
            OctopusVisualState.ASKING -> 1f + sin(time * 1.2f) * 0.02f
            OctopusVisualState.SLEEPING -> 0.95f
        }

        // Gradient colors with working pulse
        val topColor = when (visualState) {
            OctopusVisualState.WORKING -> {
                val t = sin(time * TerrariumTiming.THINKING_PULSE_SPEED) * 0.5f + 0.5f
                lerpColor(LAVENDER_TOP, LAVENDER_BRIGHT, t)
            }
            OctopusVisualState.SLEEPING -> LAVENDER_DIM
            else -> LAVENDER_TOP
        }
        val bottomColor = when (visualState) {
            OctopusVisualState.SLEEPING -> DEEP_BLUE_DIM
            else -> DEEP_BLUE_BOTTOM
        }

        val gradient = Brush.linearGradient(
            colors = listOf(topColor, bottomColor),
            start = Offset.Zero,
            end = Offset(CreatureGeometry.CODEX_VIEWBOX, CreatureGeometry.CODEX_VIEWBOX),
        )
        val markSize = bodyRadius * 1.55f * breathScale
        val markScale = markSize / CreatureGeometry.CODEX_VIEWBOX
        scope.withTransform({
            translate(cx - markSize / 2f, cy - markSize / 2f)
            scale(markScale, markScale, pivot = Offset.Zero)
        }) {
            drawPath(codexPath, brush = gradient, alpha = alpha)
        }
    }

    /**
     * Bioluminescent orbiting particles — shown when WORKING.
     * 8 particles orbit the cloud body at varying speeds/radii.
     */
    private fun drawGlowParticles(
        scope: DrawScope,
        cx: Float, cy: Float,
        bodyRadius: Float,
        alpha: Float,
    ) {
        for (particle in glowParticles) {
            val angle = time * particle.orbitSpeed + particle.phaseOffset
            val orbitR = bodyRadius * particle.orbitRadius
            val px = cx + cos(angle) * orbitR
            val py = cy + sin(angle) * orbitR * 0.7f // oval orbit

            val particleSize = bodyRadius * particle.size
            val pulseAlpha = (sin(time * 3f + particle.phaseOffset) * 0.3f + 0.7f) * alpha * 0.6f

            // Outer glow
            scope.drawCircle(
                color = particle.color,
                alpha = pulseAlpha * 0.3f,
                radius = particleSize * 2.5f,
                center = Offset(px, py),
            )
            // Core
            scope.drawCircle(
                color = Color.White,
                alpha = pulseAlpha * 0.8f,
                radius = particleSize,
                center = Offset(px, py),
            )
        }
    }

    // Pre-allocated speech bubble tail Path
    private val bubbleTailPath = Path()

    /** Speech bubble with "?" — shown during ASKING state. */
    private fun drawSpeechBubble(scope: DrawScope, cx: Float, cy: Float, bodyRadius: Float) {
        val bubbleX = cx + bodyRadius * 1.3f
        val bubbleY = cy
        val bubbleR = bodyRadius * 0.7f

        val pulse = sin(time * 2.5f) * 0.08f + 1f
        val r = bubbleR * pulse

        scope.drawCircle(
            color = Color.White,
            alpha = 0.25f,
            radius = r,
            center = Offset(bubbleX, bubbleY),
        )
        scope.drawCircle(
            color = TerrariumColors.HUDText,
            alpha = 0.5f,
            radius = r,
            center = Offset(bubbleX, bubbleY),
            style = Stroke(width = bodyRadius * 0.04f),
        )

        bubbleTailPath.reset()
        bubbleTailPath.moveTo(bubbleX - r * 0.3f, bubbleY + r * 0.3f)
        bubbleTailPath.lineTo(cx + bodyRadius * 0.6f, cy)
        bubbleTailPath.lineTo(bubbleX - r * 0.05f, bubbleY + r * 0.5f)
        bubbleTailPath.close()
        scope.drawPath(bubbleTailPath, color = Color.White, alpha = 0.25f)

        val canvas = scope.drawContext.canvas.nativeCanvas
        val textSize = r * 1.2f
        canvas.drawText(
            "?", bubbleX, bubbleY + textSize * 0.35f,
            questionMarkPaint.apply { this.textSize = textSize },
        )
    }

    // Cached name tag TEXT layout — avoids per-frame measureText.
    // Position (tagBottomY) NOT cached — depends on creature's live Y.
    private var cachedNameLayout: CachedNameLayout? = null
    private data class CachedNameLayout(
        val name: String, val bodyMetric: Float,
        val lines: List<String>, val lineHeight: Float,
        val tagWidth: Float, val tagHeight: Float, val fontSize: Float,
    )

    /** Name tag above the cloud — only shown in multi-session mode. */
    private fun drawNameTag(scope: DrawScope, cx: Float, cy: Float, bodyRadius: Float, name: String) {
        val bodyMetric = creatureNameTagMetric(scope.size.width, scaleFactor)
        val bodyTopY = cy - bodyRadius * 0.60f
        val tagBottomY = bodyTopY - bodyMetric * CreatureNameTagStyle.GAP_RATIO

        val cached = cachedNameLayout
        val tagWidth: Float
        val tagHeight: Float
        val chosenSize: Float
        val lines: List<String>
        val lineHeight: Float

        if (cached != null && cached.name == name && cached.bodyMetric == bodyMetric) {
            tagWidth = cached.tagWidth
            tagHeight = cached.tagHeight
            chosenSize = cached.fontSize
            lines = cached.lines
            lineHeight = cached.lineHeight
        } else {
            val layout = resolveCreatureNameTagLayout(
                name = name,
                bodyTopY = bodyTopY,
                bodyMetric = bodyMetric,
                paint = nameTagPaint,
            )
            tagWidth = layout.tagWidth
            tagHeight = layout.tagHeight
            chosenSize = layout.fontSize
            lines = layout.lines
            lineHeight = layout.lineHeight
            cachedNameLayout = CachedNameLayout(
                name = name,
                bodyMetric = bodyMetric,
                lines = lines,
                lineHeight = lineHeight,
                tagWidth = tagWidth,
                tagHeight = tagHeight,
                fontSize = chosenSize,
            )
        }

        val canvas = scope.drawContext.canvas.nativeCanvas

        // Hat background — codex blue-purple tint
        scope.drawRoundRect(
            color = DEEP_BLUE_BOTTOM,
            alpha = 0.6f,
            topLeft = Offset(cx - tagWidth / 2, tagBottomY - tagHeight),
            size = Size(tagWidth, tagHeight),
            cornerRadius = CornerRadius(4f, 4f),
        )

        nameTagPaint.textSize = chosenSize
        if (lines.size == 1) {
            canvas.drawText(
                lines[0], cx, tagBottomY - tagHeight * 0.25f,
                nameTagPaint,
            )
        } else {
            val topY = tagBottomY - tagHeight + chosenSize * 0.3f + chosenSize
            for (i in lines.indices) {
                canvas.drawText(
                    lines[i], cx, topY + i * lineHeight,
                    nameTagPaint,
                )
            }
        }
    }

    private fun lerpColor(a: Color, b: Color, t: Float): Color {
        return Color(
            red = a.red + (b.red - a.red) * t,
            green = a.green + (b.green - a.green) * t,
            blue = a.blue + (b.blue - a.blue) * t,
            alpha = a.alpha + (b.alpha - a.alpha) * t,
        )
    }

    private val questionMarkPaint = Paint().apply {
        isAntiAlias = true
        color = android.graphics.Color.argb(180, 226, 232, 240)
        textAlign = Paint.Align.CENTER
        typeface = Typeface.DEFAULT_BOLD
    }

    private val nameTagPaint = Paint().apply {
        isAntiAlias = true
        color = android.graphics.Color.argb(220, 226, 232, 240)
        textAlign = Paint.Align.CENTER
        typeface = Typeface.create("sans-serif", Typeface.NORMAL)
    }

    companion object {
        // Positions
        /** FLOATING (idle): rests on floor. */
        private const val STANDING_Y = 0.635f
        /** SLEEPING: dim, near bottom. */
        private const val STANDING_Y_DEEP = 0.75f
        /** ASKING: mid-water. */
        private const val ASKING_Y = 0.45f
        /** WORKING: floats near surface, wide drift zone. */
        private const val WORKING_CENTER_Y = 0.10f
        private const val WORKING_MIN_Y = 0.05f
        private const val WORKING_MAX_Y = 0.25f

        // Gradient colors — from official codex-color.svg: #B1A7FF → #7A9DFF → #3941FF
        private val LAVENDER_TOP = Color(0xFFB1A7FF)       // lavender (official top)
        private val DEEP_BLUE_BOTTOM = Color(0xFF3941FF)   // vivid blue (official bottom)
        private val LAVENDER_BRIGHT = Color(0xFFD0C4FF)    // working pulse bright
        private val LAVENDER_DIM = Color(0xFF7A7099)       // sleeping top
        private val DEEP_BLUE_DIM = Color(0xFF2A2F80)      // sleeping bottom

        // Bioluminescent glow particle colors
        private val GLOW_CYAN = Color(0xFF66DDFF)
        private val GLOW_LAVENDER = Color(0xFFD0AAFF)
        private const val GLOW_PARTICLE_COUNT = 8

        // Normalise arc-flag compression before parsing — Compose's PathParser
        // is mostly tolerant of SVG flag compression but the core graphics
        // parser that backs the e-ink path is not. Normalising here keeps the
        // two surfaces byte-for-byte equivalent in input shape.
        private val codexPath = PathParser()
            .parsePathString(normalizeSvgArcFlags(CreatureGeometry.CODEX_PATH_DATA))
            .toPath().apply {
                fillType = PathFillType.EvenOdd
            }
    }
}

File: apple/AgentDeck/Terrarium/Creatures/CloudCreature.swift
// CloudCreature.swift — Cloud creature for Codex CLI
// Exact design/brand/codex.svg geometry with a lavender → blue gradient.

import SwiftUI

// MARK: - Cloud Visual State

enum CloudVisualState {
    case dormant
    case drifting   // Idle — gentle pulse, slow drift
    case pulsing    // Processing — vivid gradient, glow, faster morph
    case waiting    // Awaiting — "?" bubble
}

final class CloudCreature: Creature {
    private static let codexPathData = "M8.086.457a6.105 6.105 0 013.046-.415c1.333.153 2.521.72 3.564 1.7a.117.117 0 00.107.029c1.408-.346 2.762-.224 4.061.366l.063.03.154.076c1.357.703 2.33 1.77 2.918 3.198.278.679.418 1.388.421 2.126a5.655 5.655 0 01-.18 1.631.167.167 0 00.04.155 5.982 5.982 0 011.578 2.891c.385 1.901-.01 3.615-1.183 5.14l-.182.22a6.063 6.063 0 01-2.934 1.851.162.162 0 00-.108.102c-.255.736-.511 1.364-.987 1.992-1.199 1.582-2.962 2.462-4.948 2.451-1.583-.008-2.986-.587-4.21-1.736a.145.145 0 00-.14-.032c-.518.167-1.04.191-1.604.185a5.924 5.924 0 01-2.595-.622 6.058 6.058 0 01-2.146-1.781c-.203-.269-.404-.522-.551-.821a7.74 7.74 0 01-.495-1.283 6.11 6.11 0 01-.017-3.064.166.166 0 00.008-.074.115.115 0 00-.037-.064 5.958 5.958 0 01-1.38-2.202 5.196 5.196 0 01-.333-1.589 6.915 6.915 0 01.188-2.132c.45-1.484 1.309-2.648 2.577-3.493.282-.188.55-.334.802-.438.286-.12.573-.22.861-.304a.129.129 0 00.087-.087A6.016 6.016 0 015.635 2.31C6.315 1.464 7.132.846 8.086.457zm-.804 7.85a.848.848 0 00-1.473.842l1.694 2.965-1.688 2.848a.849.849 0 001.46.864l1.94-3.272a.849.849 0 00.007-.854l-1.94-3.393zm5.446 6.24a.849.849 0 000 1.695h4.848a.849.849 0 000-1.696h-4.848z"
    private static let codexPath = CrayfishCreature.parseSvgPath(codexPathData)

    // MARK: - Properties

    let sessionId: String
    var displayName: String?
    var visualState: CloudVisualState = .drifting
    var homeX: Float
    var homeY: Float
    var scale: Float

    private var time: Float = 0
    private(set) var currentX: Float
    private(set) var currentY: Float
    private var phaseOffset: Float
    private var driftPhase: Float

    private var previousState: CloudVisualState?
    private var transitionProgress: Float = 1.0

    var onWaitingExit: (() -> Void)?

    // MARK: - Init

    init(sessionId: String, homeX: Float, homeY: Float, scale: Float) {
        self.sessionId = sessionId
        self.homeX = homeX
        self.homeY = homeY
        self.scale = scale
        self.currentX = homeX
        self.currentY = homeY
        self.phaseOffset = Float.random(in: 0...Float.pi * 2)
        self.driftPhase = Float.random(in: 0...Float.pi * 2)
    }

    // MARK: - Update

    func update(dt: Float, state: TerrariumState) {
        time += dt

        if let creature = state.cloudCreatures.first(where: { $0.id == sessionId }) {
            let newState = creature.state
            if newState != visualState {
                if visualState == .waiting { onWaitingExit?() }
                previousState = visualState
                transitionProgress = 0
                visualState = newState
            }
        }

        if transitionProgress < 1.0 {
            transitionProgress = min(1.0, transitionProgress + dt * 2.5)
        }

        updatePosition(dt: dt)
    }

    private func updatePosition(dt: Float) {
        let idleY = min(0.62, max(0.58, homeY + 0.30))
        let waitingY = min(0.54, max(0.48, homeY + 0.18))
        let pulseY = min(0.25, max(0.05, homeY))
        let targetY: Float = switch visualState {
        case .dormant: 0.70
        case .drifting: idleY
        case .pulsing: pulseY
        case .waiting: waitingY
        }

        let lerpRate: Float = visualState == .pulsing ? 2.0 : 1.5
        let pulseSpeed: Float = visualState == .pulsing ? 1.5 : 0.5
        let pulseAmp: Float = visualState == .pulsing ? 0.015 : 0.008
        let pulseBob = sin((time + phaseOffset) * pulseSpeed) * pulseAmp
        currentY += (targetY + pulseBob - currentY) * dt * lerpRate

        // Processing: wider horizontal drift (floating near surface, drifting side to side)
        let driftAmp: Float = visualState == .pulsing ? min(0.05, 0.02 + scale * 0.03) : 0.006
        let driftSpeed: Float = visualState == .pulsing ? 0.15 : 0.3
        let driftX = sin((time + driftPhase) * driftSpeed) * driftAmp
        currentX += (homeX + driftX - currentX) * dt * lerpRate

        let minX = max(0.18, homeX - 0.08)
        let maxX = min(0.72, homeX + 0.08)
        currentX = min(maxX, max(minX, currentX))
        currentY = min(0.62, max(0.08, currentY))
    }

    func currentPosition() -> (x: Float, y: Float) { (currentX, currentY) }
    func isPulsing() -> Bool { visualState == .pulsing }

    // MARK: - Draw

    func draw(context: inout GraphicsContext, size: CGSize) {
        guard visualState != .dormant else { return }

        let w = Float(size.width)
        let h = Float(size.height)
        let bodyWidth = w * 0.060 * scale

        let cx = CGFloat(currentX * w)
        let bobOffset = visualState == .pulsing ? CGFloat(sin(time * 2.0) * h * 0.008) : 0
        let cy = CGFloat(currentY * h) + bobOffset

        // Glow behind cloud (processing)
        if visualState == .pulsing {
            drawGlow(context: &context, cx: cx, cy: cy, bodyW: CGFloat(bodyWidth))
        }

        // Canonical Codex mark
        drawCloudBody(context: &context, cx: cx, cy: cy, bodyW: CGFloat(bodyWidth))

        // "?" bubble
        if visualState == .waiting {
            drawSpeechBubble(context: &context, cx: cx, cy: cy, bodyW: CGFloat(bodyWidth))
        }

        // Name tag
        if let name = displayName {
            drawNameTag(
                context: &context,
                name: name,
                cx: cx,
                cy: cy,
                bodyW: CGFloat(bodyWidth),
                canvasWidth: size.width
            )
        }
    }

    // MARK: - Canonical Codex mark

    private func drawCloudBody(context: inout GraphicsContext, cx: CGFloat, cy: CGFloat, bodyW: CGFloat) {
        let alpha = visualState == .dormant ? 0.3 : 0.9
        let breathScale: CGFloat = 1.0 + CGFloat(sin(time * (visualState == .pulsing ? 2.0 : 0.6)) * 0.03)

        // Gradient colors
        let topColor: Color
        let bottomColor: Color
        if visualState == .pulsing {
            let pulse = sin(time * 2.5) * 0.3 + 0.7
            topColor = TerrariumColors.lerpColor(
                TerrariumColors.cloudHighlight, Color.white, Float(pulse) * 0.2)
            bottomColor = TerrariumColors.lerpColor(
                TerrariumColors.cloudDeep, TerrariumColors.cloudBell, Float(pulse) * 0.3)
        } else {
            topColor = TerrariumColors.cloudHighlight
            bottomColor = TerrariumColors.cloudDeep
        }

        let markSize = bodyW * 1.28 * breathScale
        let markScale = markSize / 24
        context.drawLayer { bodyCtx in
            bodyCtx.opacity = alpha
            bodyCtx.translateBy(x: cx - markSize / 2, y: cy - markSize / 2)
            bodyCtx.scaleBy(x: markScale, y: markScale)
            bodyCtx.fill(Self.codexPath, with: .linearGradient(
                Gradient(colors: [topColor, bottomColor]),
                startPoint: .zero,
                endPoint: CGPoint(x: 24, y: 24)
            ), style: FillStyle(eoFill: true))
        }
    }



    // MARK: - Glow

    private func drawGlow(context: inout GraphicsContext, cx: CGFloat, cy: CGFloat, bodyW: CGFloat) {
        let glowPulse = CGFloat(sin(time * 2.5) * 0.3 + 0.7)
        let glowR = bodyW * 1.0 * glowPulse

        let rect = CGRect(x: cx - glowR, y: cy - glowR, width: glowR * 2, height: glowR * 2)
        context.fill(Path(ellipseIn: rect),
                     with: .color(TerrariumColors.cloudGlow.opacity(0.1 * Double(glowPulse))))

        for i in 0..<4 {
            let angle = CGFloat(Float(i) / 4 * Float.pi * 2 + time * 0.5)
            let orbitR = bodyW * 0.65
            let px = cx + cos(angle) * orbitR
            let py = cy + sin(angle) * orbitR * 0.6
            let pSize = bodyW * 0.03
            let pRect = CGRect(x: px - pSize, y: py - pSize, width: pSize * 2, height: pSize * 2)
            context.fill(Path(ellipseIn: pRect),
                         with: .color(TerrariumColors.cloudGlow.opacity(0.3 * Double(glowPulse))))
        }
    }

    // MARK: - Speech Bubble

    private func drawSpeechBubble(context: inout GraphicsContext, cx: CGFloat, cy: CGFloat, bodyW: CGFloat) {
        let bx = cx + bodyW * 0.55
        let by = cy - bodyW * 0.1
        let br = bodyW * 0.22
        let pulse = CGFloat(sin(time * 2.5)) * 0.08 + 1
        let r = br * pulse

        let rect = CGRect(x: bx - r, y: by - r, width: r * 2, height: r * 2)
        context.fill(Path(ellipseIn: rect), with: .color(.white.opacity(0.25)))
        context.stroke(Path(ellipseIn: rect),
                       with: .color(TerrariumColors.hudText.opacity(0.5)),
                       lineWidth: bodyW * 0.02)

        var tail = SwiftUI.Path()
        tail.move(to: CGPoint(x: bx - r * 0.3, y: by + r * 0.3))
        tail.addLine(to: CGPoint(x: cx + bodyW * 0.28, y: cy))
        tail.addLine(to: CGPoint(x: bx - r * 0.05, y: by + r * 0.5))
        tail.closeSubpath()
        context.fill(tail, with: .color(.white.opacity(0.25)))

        context.draw(
            Text("?").font(.system(size: r * 1.2, weight: .bold)).foregroundColor(TerrariumColors.hudText.opacity(0.7)),
            at: CGPoint(x: bx, y: by)
        )
    }

    // MARK: - Name Tag

    private func drawNameTag(context: inout GraphicsContext, name: String,
                             cx: CGFloat, cy: CGFloat, bodyW: CGFloat, canvasWidth: CGFloat) {
        drawTerrariumNameTag(
            context: &context,
            name: name,
            cx: cx,
            bodyTopY: cy - bodyW * 0.6,
            bodyMetric: terrariumNameTagMetric(canvasWidth: canvasWidth, scale: scale),
            backgroundColor: TerrariumColors.cloudNameBg
        )
    }
}
